presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java

```diff
@@ -374,6 +385,15 @@ public static QueryStats create(
 
             completeInfo = completeInfo && stageInfo.isFinalStageInfo();
             operatorStatsSummary.addAll(stageExecutionStats.getOperatorSummaries());
+            if (stageExecutionStats.getRuntimeStats() != null) {
```

Comment:
We don't usually pass nulls around in our codebase. Let's always assume we can get something

Reply:
This also has similar binary incompatible issue. If the coordinator is running a new binary version and the worker is running an old binary version, this could be null in the coordinator. Do we have this case during release? Should we allow this to be null?

Reply:
> Failure seems related:
> 
> ```
> Error:  testGetInfoForQuery(com.facebook.presto.tests.querystats.TestHttpQueryStatsClient)  Time elapsed: 0.226 s  <<< FAILURE!
> java.io.UncheckedIOException: 
> com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `com.facebook.presto.execution.QueryStats`, problem: runtimeStats is null
>  at [Source: UNKNOWN; line: -1, column: -1]
> 	at com.facebook.presto.tests.querystats.HttpQueryStatsClient$GetQueryStatsResponseHandler.handle(HttpQueryStatsClient.java:88)
> 	at com.facebook.presto.tests.querystats.HttpQueryStatsClient$GetQueryStatsResponseHandler.handle(HttpQueryStatsClient.java:59)
> 	at com.facebook.airlift.http.client.testing.TestingHttpClient.execute(TestingHttpClient.java:111)
> 	at com.facebook.airlift.http.client.testing.TestingHttpClient.execute(TestingHttpClient.java:64)
> 	at com.facebook.presto.tests.querystats.HttpQueryStatsClient.getQueryStats(HttpQueryStatsClient.java:56)
> 	at com.facebook.presto.tests.querystats.TestHttpQueryStatsClient.testGetInfoForQuery(TestHttpQueryStatsClient.java:68)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
> 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
> 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
> 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
> 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
> 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> Caused by: com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `com.facebook.presto.execution.QueryStats`, problem: runtimeStats is null
>  at [Source: UNKNOWN; line: -1, column: -1]
> 	at com.fasterxml.jackson.databind.exc.ValueInstantiationException.from(ValueInstantiationException.java:47)
> 	at com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1732)
> 	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:491)
> 	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:514)
> 	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:285)
> 	at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromObjectWith(ValueInstantiator.java:229)
> 	at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:198)
> 	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:488)
> 	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287)
> 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326)
> 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159)
> 	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4173)
> 	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2467)
> 	at com.fasterxml.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:2920)
> 	at com.facebook.presto.tests.querystats.HttpQueryStatsClient$GetQueryStatsResponseHandler.handle(HttpQueryStatsClient.java:84)
> 	... 18 more
> Caused by: java.lang.NullPointerException: runtimeStats is null
> 	at java.util.Objects.requireNonNull(Objects.java:228)
> 	at com.facebook.presto.execution.QueryStats.<init>(QueryStats.java:271)
> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
> 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
> 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
> 	at com.fasterxml.jackson.databind.introspect.AnnotatedConstructor.call(AnnotatedConstructor.java:124)
> 	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:283)
> 	... 28 more
> ```

Yes, I also found this yesterday. The runtimeStats could be null if the QueryStats is deserialized from a json generated by an older binary version. I updated the logic to create a new object in this case. Please see if it makes sense.